### PR TITLE
fix Issue 21772 - [REG2.069] Consecutive different-signed double.nans…

### PR DESCRIPTION
**Pull request by WalterBright**
… in an array literal take the sign of the previous nan (same for float and real)

### CODE REVIEW

**Review by dkorpel — COMMENTED** [review]
Bot says regressions should target stable

File: compiler/src/dmd/e2ir.d [hunk 1]
```diff
@@ -17,6 +17,7 @@ import core.stdc.string;
 import core.stdc.time;
 
 import dmd.root.array;
+import dmd.root.complex;
```
Comment (dkorpel) [inline compiler/src/dmd/e2ir.d:20]: Why is this import added?

Comment (WalterBright) [reply compiler/src/dmd/e2ir.d]: an older version of this PR

File: compiler/src/dmd/constfold.d [hunk 2]
```diff
@@ -826,6 +826,7 @@ UnionExp Equal(EXP op, const ref Loc loc, Type type, Expression e1, Expression e
 
 UnionExp Identity(EXP op, const ref Loc loc, Type type, Expression e1, Expression e2)
 {
+printf("t1: %s, t2: %s\n", e1.type.toChars(), e2.type.toChars());
```
Comment (dkorpel) [inline compiler/src/dmd/constfold.d:829]: debug statement

**Comment by ibuclaw** [comment]
Testing all versions of gdc from 9..12, I don't get the same bug as dmd here, CTFE values look fine too.
```
$ gdc r.d -o gdc-r
[-QNaN, +QNaN, +QNaN, 1.0e+0, +QNaN, -QNaN]
$ ./gdc-r
[-nan, nan, nan, 1, nan, -nan]
```
vs. dmd
```
$ dmd r.d -o dmd-r
[-nan, nan, nan, 1.0, nan, -nan]
$ ./dmd-r
[-nan, -nan, -nan, 1, nan, nan]
core.exception.AssertError@r.d(11): Assertion failure
```
So I'm not entirely sure if the CTFE changes are required here for fixing this bug.

Also bearing in mind that CTFE `nan !is -nan` was fixed in https://github.com/dlang/dmd/pull/13780.

**Review by dkorpel — APPROVED** [review]
As far as I understand, the issue is not actually a regression. It was always wrong, `writlen` just didn't print -nan before 2.069
